### PR TITLE
implement an addon using helm chart

### PR DESCRIPTION
**Pull request by zhiweiyin318**
1. implement a factory to build addonagent by helm chart.
2. can override and customize the chart values by the annotation of manangedClusterAddon CR.

limitations:
1. only support build-in object: Release.Name, Release.Namespace and Capabilities.KubeVersion.Version
2. do not support  hooks currently.

Signed-off-by: Zhiwei Yin <zyin@redhat.com>
issue : https://github.com/open-cluster-management-io/community/issues/117

### CODE REVIEW

File: pkg/addonfactory/chartaddonfactory/chartaddonagent.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+const annotationValuesName string = "addon.open-cluster-management.io/values"
+
+// the reserved fields for template config
+const (
+	clusterName           string = "clusterName"
+	addonInstallNamespace string = "addonInstallNamespace"
+	hubKubeConfigSecret   string = "hubKubeConfigSecret"
+)
+
+type ChartAgentAddon struct {
```
Comment (qiujian16) [inline pkg/addonfactory/chartaddonfactory/chartaddonagent.go]: helmagent?

File: pkg/addonfactory/chartaddonfactory/chartaddonagent.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	installNamespace := addon.Spec.InstallNamespace
+	if len(installNamespace) == 0 {
+		installNamespace = AddonDefaultInstallNamespace
+	}
+	overrideValues[addonInstallNamespace] = installNamespace
+	overrideValues[hubKubeConfigSecret] = fmt.Sprintf("%s-hub-kubeconfig", a.agentAddonOptions.AddonName)
+
+	// users can override the template config fields by the annotation of addon cr.
+	// the key of the annotation is `templateConfig`, the value is a json string which has the config struct.
+	// for example: "addon.open-cluster-management.io/values": `{"NodeSelector":{"host":"ssd"},"Image":"quay.io/helloworld:2.4"}`
```
Comment (qiujian16) [inline pkg/addonfactory/chartaddonfactory/chartaddonagent.go]: helmchart-values

File: pkg/addonfactory/chartaddonfactory/chartaddonagent.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	addon *addonapiv1alpha1.ManagedClusterAddOn) (map[string]interface{}, error) {
+	overrideValues := structs.Map(a.values)
+	overrideValues[clusterName] = cluster.GetName()
+
+	installNamespace := addon.Spec.InstallNamespace
+	if len(installNamespace) == 0 {
+		installNamespace = AddonDefaultInstallNamespace
+	}
+	overrideValues[addonInstallNamespace] = installNamespace
+	overrideValues[hubKubeConfigSecret] = fmt.Sprintf("%s-hub-kubeconfig", a.agentAddonOptions.AddonName)
```
Comment (qiujian16) [inline pkg/addonfactory/chartaddonfactory/chartaddonagent.go]: I would suggest having a separate func for builtin values. And hubKubeConfigSecret needs to be defined based on addon status

File: pkg/addonfactory/chartaddonfactory/chartaddonagent.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+	if len(annotations[annotationValuesName]) != 0 {
+		annotationValues := map[string]interface{}{}
+		err := json.Unmarshal([]byte(annotations[annotationValuesName]), &annotationValues)
+		if err != nil {
+			klog.Error("failed to unmarshal the values from annotation of addon cr %v. err:%v", annotationValues, err)
+			return nil, err
+		}
+
+		for k, v := range annotationValues {
+			overrideValues[k] = v
```
Comment (qiujian16) [inline pkg/addonfactory/chartaddonfactory/chartaddonagent.go]: this is suspicious, what if user define addonInstallNamespace in the values. Which should be overridden

File: pkg/addonfactory/chartaddonfactory/testchart/templates/clusterclaim.yaml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+apiVersion: cluster.open-cluster-management.io/v1alpha1
+kind: ClusterClaim
+metadata:
+  labels:
+    open-cluster-management.io/hub-managed: ""
+    open-cluster-management.io/spoke-only: ""
+  name: clusterName
+spec:
+  value:  {{ .Values.clusterName }}
```
Comment (qiujian16) [inline pkg/addonfactory/chartaddonfactory/testchart/templates/clusterclaim.yaml]: eof

File: pkg/addonfactory/chartaddonfactory/chartaddonagent.go (hunk 1 above)
Comment (zhiweiyin318) [reply pkg/addonfactory/chartaddonfactory/chartaddonagent.go]: defined with  `HelmAgentAddon`

File: pkg/addonfactory/chartaddonfactory/chartaddonagent.go (hunk 3 above)
Comment (zhiweiyin318) [reply pkg/addonfactory/chartaddonfactory/chartaddonagent.go]: use `getBuildInValues` to get all of build-in values. 
define `hubKubeConfigSecret ` when `a.agentAddonOptions.Registration != nil`

File: pkg/addonfactory/chartaddonfactory/chartaddonagent.go (hunk 4 above)
Comment (zhiweiyin318) [reply pkg/addonfactory/chartaddonfactory/chartaddonagent.go]: I define `Values map[string]interface{}` to replace the `Values`  defined in chart lib.
user needs to define  
```
getValuesFunc func(cluster *clusterv1.ManagedCluster,
	addon *addonapiv1alpha1.ManagedClusterAddOn) (Values, error). 
```

and can use `GetValuesFromAddonAnnotation` to get values in annotation of addon CR, and use `MergeValues` to merge 2 Values.

File: pkg/addonfactory/helmaddonfactory/helmagentaddon.go [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+	buildInValues[clusterName] = cluster.GetName()
+
+	installNamespace := addon.Spec.InstallNamespace
+	if len(installNamespace) == 0 {
+		installNamespace = AddonDefaultInstallNamespace
+	}
+	buildInValues[addonInstallNamespace] = installNamespace
+
+	if a.agentAddonOptions.Registration != nil {
```
Comment (qiujian16) [inline pkg/addonfactory/helmaddonfactory/helmagentaddon.go]: Add a comment and todo here. It also depdends on the signer configuration in registration, and the registration is an array actually

**Review by qiujian16 — COMMENTED** [review]
/approve

I think we need an e2e test as the followup

/assign @yue9944882

File: examples/helmAgentAddon/example/policyaddon/chart/templates/policy.open-cluster-management.io_configurationpolicies_crd.yaml [hunk 7]
```diff
@@ -0,0 +1,387 @@
+{{- if semverCompare "< 1.16.0" .Capabilities.KubeVersion.Version }}
+apiVersion: apiextensions.k8s.io/v1beta1
```
Comment (yue9944882) [inline examples/helmAgentAddon/example/policyaddon/chart/templates/policy.open-cluster-management.io_configurationpolicies_crd.yaml]: i think CRDs are expected to be placed in a dedicated folder named `crds`, is this supported?

File: examples/helmAgentAddon/example/policyaddon/main.go [hunk 8]
```diff
@@ -0,0 +1,143 @@
+package main
```
Comment (yue9944882) [inline examples/helmAgentAddon/example/policyaddon/main.go]: > examples/helmAgentAddon/example/policyaddon/main.go 

nit: conventionally the folder/file name should be snakecased?

File: examples/helmAgentAddon/helmAgentAddon.md [hunk 9]
```diff
@@ -0,0 +1,48 @@
+# Overview
```
Comment (yue9944882) [inline examples/helmAgentAddon/helmAgentAddon.md]: ditto

File: pkg/addonfactory/helmaddonfactory/helmagentaddon.go (hunk 6 above)
Comment (zhiweiyin318) [reply pkg/addonfactory/helmaddonfactory/helmagentaddon.go]: added a TODO here

File: examples/helmAgentAddon/example/policyaddon/chart/templates/policy.open-cluster-management.io_configurationpolicies_crd.yaml (hunk 7 above)
Comment (zhiweiyin318) [reply examples/helmAgentAddon/example/policyaddon/chart/templates/policy.open-cluster-management.io_configurationpolicies_crd.yaml]: I missed this feature. have supported it . and added ut for this case. thanks

File: examples/helmAgentAddon/example/policyaddon/main.go (hunk 8 above)
Comment (zhiweiyin318) [reply examples/helmAgentAddon/example/policyaddon/main.go]: done

File: examples/helmAgentAddon/helmAgentAddon.md (hunk 9 above)
Comment (zhiweiyin318) [reply examples/helmAgentAddon/helmAgentAddon.md]: done

**Comment by zhiweiyin318** [comment]
added an example `helloworld_helm` in examples, and use the example for e2e test.

File: docs/helmAgentAddon.md [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+* Release.Name
+* Release.Namespace
+
+2. Not support Hooks of Helm Chart currently.
+
+## Migration 
+We have an example for Helm Chart migration in [helloworld_helm](../examples/helloworld_helm).
+1. Copy your helm charts into repo.
+2. Create and start helmAgentAddon instance like this:
+   ```goZ
```
Comment (qiujian16) [inline docs/helmAgentAddon.md]: rm Z

File: docs/helmAgentAddon.md [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+1. Not all of build-in Objects of Helm Chart are support in Addon. We only support below:
+* Capabilities.KubeVersion
+* Release.Name
+* Release.Namespace
+
+2. Not support Hooks of Helm Chart currently.
+
+## Migration 
+We have an example for Helm Chart migration in [helloworld_helm](../examples/helloworld_helm).
+1. Copy your helm charts into repo.
```
Comment (qiujian16) [inline docs/helmAgentAddon.md]: I think you would need to import this into embed.FS

File: examples/helloworld_helm/helloworld_helm_agent.go [hunk 12]
```diff
@@ -0,0 +1,157 @@
+package main
+
```
Comment (qiujian16) [inline examples/helloworld_helm/helloworld_helm_agent.go:2]: can we just build a common helloworld agent? It seems each example shows the same agent.

File: pkg/addonfactory/helmaddonfactory/helm_agentaddon.go [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+		a.releaseOptions(cluster, addon), a.capabilities(cluster, addon))
+	if err != nil {
+		klog.Error("failed to render helm chart with values %v. err:%v", overrideValues, err)
+		return values, err
+	}
+
+	return values, nil
+}
+
+func (a *HelmAgentAddon) getBuildInValues(
```
Comment (qiujian16) [inline pkg/addonfactory/helmaddonfactory/helm_agentaddon.go]: BuiltIn

File: pkg/addonfactory/helmaddonfactory/helm_addonfactory.go [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+		}
+		bfs = append(bfs, bf)
+	}
+
+	userChart, err := loader.LoadFiles(bfs)
+	if err != nil {
+		klog.Errorf("failed to load chart. err:%v", err)
+		return nil, err
+	}
+	klog.Info(userChart.ChartPath())
```
Comment (qiujian16) [inline pkg/addonfactory/helmaddonfactory/helm_addonfactory.go]: do you need these Info?

**Review by qiujian16 — COMMENTED** [review]
/approve

some nit comments

File: docs/helmAgentAddon.md (hunk 10 above)
Comment (zhiweiyin318) [reply docs/helmAgentAddon.md]: done

File: docs/helmAgentAddon.md (hunk 11 above)
Comment (zhiweiyin318) [reply docs/helmAgentAddon.md]: done

File: pkg/addonfactory/helmaddonfactory/helm_agentaddon.go (hunk 13 above)
Comment (zhiweiyin318) [reply pkg/addonfactory/helmaddonfactory/helm_agentaddon.go]: done

File: pkg/addonfactory/helmaddonfactory/helm_addonfactory.go (hunk 14 above)
Comment (zhiweiyin318) [reply pkg/addonfactory/helmaddonfactory/helm_addonfactory.go]: deleted.

File: examples/helloworld_helm/helloworld_helm_agent.go (hunk 12 above)
Comment (zhiweiyin318) [reply examples/helloworld_helm/helloworld_helm_agent.go:2]: yes, the same agent in each examples.  let me follow up to refactor the examples in another template PR.

**Pull request merged**